data/txHandler.go

```diff
@@ -286,37 +313,110 @@ func (handler *TxHandler) asyncVerifySignature(arg interface{}) interface{} {
+			return true
+		}
+	} else {
+		// a transaction group => cache/dedup the entire group canonical group
+		encodeBuf := make([]byte, 0, unverifiedTxGroup[0].Msgsize()*ntx)
+		for i := range unverifiedTxGroup {
+			encodeBuf = unverifiedTxGroup[i].MarshalMsg(encodeBuf)
+		}
+		if len(encodeBuf) != consumed {
+			// reallocated, some assumption on size was wrong
+			// log and skip
+			logging.Base().Warnf("Decoded size %d does not match to encoded %d", consumed, len(encodeBuf))
+		} else {
+			d := crypto.Hash(encodeBuf)
+			if handler.txCanonicalCache.CheckAndPut(&d) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// processIncomingTxn decodes a transaction group from incoming message and enqueues into the back log for processing.
+// The function also performs some input data pre-validation;
+//  - txn groups are cut to MaxTxGroupSize size
+//  - message are checked for duplicates
+//  - transactions are checked for duplicates
 func (handler *TxHandler) processIncomingTxn(rawmsg network.IncomingMessage) network.OutgoingMessage {
+
+	if handler.cacheConfig.enableFilteringRawMsg {
+		// check for duplicate messages
+		// this helps against relaying duplicates
+		if handler.msgCache.CheckAndPut(rawmsg.Data) {
+			transactionMessagesDupRawMsg.Inc(nil)
+			return network.OutgoingMessage{Action: network.Ignore}
+		}
+	}
+
+	unverifiedTxGroup := make([]transactions.SignedTxn, 1)
 	dec := protocol.NewMsgpDecoderBytes(rawmsg.Data)
 	ntx := 0
-	unverifiedTxGroup := make([]transactions.SignedTxn, 1)
+	consumed := 0
 	for {
 		if len(unverifiedTxGroup) == ntx {
 			n := make([]transactions.SignedTxn, len(unverifiedTxGroup)*2)
 			copy(n, unverifiedTxGroup)
 			unverifiedTxGroup = n
 		}
-
 		err := dec.Decode(&unverifiedTxGroup[ntx])
-		if err == io.EOF {
-			break
-		}
 		if err != nil {
+			if err == io.EOF {
+				break
+			}
 			logging.Base().Warnf("Received a non-decodable txn: %v", err)
 			return network.OutgoingMessage{Action: network.Disconnect}
 		}
+		consumed = dec.Consumed()
 		ntx++
+		if ntx >= config.MaxTxGroupSize {
+			// max ever possible group size reached, done reading input.
+			if dec.Remaining() > 0 {
+				// if something else left in the buffer - this is an error, drop
+				transactionMessageTxGroupExcessive.Inc(nil)
+				return network.OutgoingMessage{Action: network.Disconnect}
+			}
+		}
 	}
 	if ntx == 0 {
 		logging.Base().Warnf("Received empty tx group")
 		return network.OutgoingMessage{Action: network.Disconnect}
 	}
+
 	unverifiedTxGroup = unverifiedTxGroup[:ntx]
 
 	if ntx == config.MaxTxGroupSize {
 		transactionMessageTxGroupFull.Inc(nil)
-	} else if ntx > config.MaxTxGroupSize {
-		transactionMessageTxGroupExcessive.Inc(nil)
+	}
+
+	if handler.cacheConfig.enableFilteringCanonical {
+		if handler.dedupCanonical(ntx, unverifiedTxGroup, consumed) {
+			transactionMessagesDupCanonical.Inc(nil)
+			return network.OutgoingMessage{Action: network.Ignore}
+		}
 	}
 
 	select {
```

Comment:
Wouldnt we want to remove the item from the dedup cache if the tx doesn't make it to the backlog queue and is dropped?